File: Interview-Preparation/MissionJune2020/src/com/interviewbit/LinkedLists/PalindromeList.java
package com.interviewbit.LinkedLists;

public class PalindromeList {

    static class ListNode {
        int val;
        ListNode next;
        ListNode(int value) {
            val = value;
            next = null;
        }
    }

    public static int getListLength(final ListNode head) {
        int length = 0;
        ListNode curr = head;
        while (curr != null) {
            length++;
            curr = curr.next;
        }
        return length;
    }

    public static void printList(final ListNode head) {
        ListNode curr = head;
        while (curr != null) {
            System.out.print(curr.val + " ");
            curr = curr.next;
        }
        System.out.println();
    }

    public static ListNode reverseList(ListNode head) {
        ListNode curr = head, prev = null, next = null;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public static int isPalindrome(ListNode head) {
        if (head == null || head.next == null) return 1;
        int length = getListLength(head);
        ListNode first = head, mid = null, last = head;
        if (length % 2 != 0) {
            int counter = length / 2;
            while (--counter > 0) {
                last = last.next;
            }
            mid = last.next;
            last.next = null;
            last = mid.next;
            last = reverseList(last);
        } else {
            int counter = length / 2;
            while (--counter > 0) {
                last = last.next;
            }
            ListNode temp = last;
            last = last.next;
            temp.next = null;
            last = reverseList(last);
        }
        while (first != null && last != null) {
            if (first.val != last.val) return 0;
            first = first.next;
            last = last.next;
        }
        return 1;
    }

    public static void main(String[] args) {

        ListNode head = new ListNode(1);
        head.next = new ListNode(2);
        head.next.next = new ListNode(1);
        head.next.next.next = new ListNode(1);
        //head.next.next.next.next = new ListNode(1);
        printList(head);
        System.out.println(isPalindrome(head));
    }
}

File: InterviewBit/20210701_2204/L4-Linked_Lists/palindrome-list.cpp
#include <iostream>

struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(NULL) {}
};

ListNode *reverse(ListNode *head) {
  ListNode *curr = head, *prev = nullptr, *next = nullptr;
  while (curr) {
    next = curr->next;
    curr->next = prev;
    prev = curr;
    curr = next;
  }
  return prev;
}

bool check(ListNode *l1, ListNode *l2) {
  while (l1 && l2) {
    if (l1->val != l2->val)
      return false;
    l1 = l1->next;
    l2 = l2->next;
  }
  return true;
}

int lPalin(ListNode *A) {
  if (!A)
    return true;
  ListNode *slow = A, *fast = A;
  while (fast->next && fast->next->next) {
    slow = slow->next;
    fast = fast->next->next;
  }
  ListNode *l1 = A;
  ListNode *l2 = reverse(slow->next);
  slow->next = nullptr;
  return check(l1, l2);
}